src/crimson/os/seastore/journal.cc

```diff
@@ -600,37 +600,50 @@ Journal::replay_ret Journal::replay(delta_handler_t &&delta_handler)
 }
 
 Journal::scan_extents_ret Journal::scan_extents(
-  paddr_t addr,
+  scan_extents_cursor &cursor,
   extent_len_t bytes_to_read)
 {
-  // Caller doesn't know addr of first record, so addr.offset == 0 is special
-  if (addr.offset == 0) addr.offset = block_size;
-
-  return read_segment_header(addr.segment
+  auto ret = std::make_unique<scan_extents_ret_bare>();
+  auto &retref = *ret;
+  return read_segment_header(cursor.get_offset().segment
   ).handle_error(
     scan_extents_ertr::pass_further{},
     crimson::ct_error::assert_all{}
-  ).safe_then([=](auto header) {
+  ).safe_then([&](auto header) {
     return seastar::do_with(
-      scan_extents_ret_bare(),
-      [=](auto &ret) {
-	return seastar::do_with(
-	  extent_handler_t([&ret](auto addr, const auto &info) mutable {
-	    ret.second.push_back(std::make_pair(addr, info));
-	    return scan_extents_ertr::now();
-	  }),
-	  [=, &ret](auto &handler) mutable {
-	    return scan_segment(
-	      addr,
-	      bytes_to_read,
-	      header.segment_nonce,
-	      nullptr,
-	      &handler).safe_then([&ret](auto next) mutable {
-		ret.first = next;
-		return std::move(ret);
-	      });
-	  });
+      found_record_handler_t(
+	[&](
+	  paddr_t base,
+	  const record_header_t &header,
+	  const bufferlist &mdbuf) mutable {
+
+	  auto infos = try_decode_extent_infos(
+	    header,
+	    mdbuf);
+	  if (!infos) {
+	    // This should be impossible, we did check the crc on the mdbuf
+	    logger().error(
+	      "Journal::scan_extents unable to decode extents for record {}",
+	      base);
+	    assert(infos);
+	  }
+
+	  paddr_t extent_offset = base.add_offset(header.mdlength);
+	  for (const auto &i : *infos) {
+	    retref.emplace_back(extent_offset, i);
+	    extent_offset.offset += i.len;
+	  }
+	  return scan_extents_ertr::now();
+	}),
+      [&](auto &dhandler) {
+	return scan_valid_records(
```

Comment:
Is it a fatal error if any record becomes invalid during `scan_extents()`, infering data corruption from a dying SSD?

Reply:
No, after a power-off event, we'd expect the tail records of any actively written segment to be invalid.

Reply:
I mean, do we need to worry about data corruption from a dying SSD in seastore?

Specifically, during `Journal::scan_extents()`, if unexpected bit flip is happened from the underlying device, causing crc mismatch, the current implementation just ignores the follow-up committed records in this segment and keeps on GC without error. Then later, a higher level application may crush due to the corrupted extents. And at that time, it could be very hard to distinguish whether it is a rear but fatal bug of the application, or it is a side-effect of data corruption.

Reply:
Yeah, that's potentially a problem.  At some point we'll probably want an independent registry of segment endpoints, but I'm leaving it for future work.

Reply:
More general checksumming in the normal read path will be via checksums embedded in the lba tree, btw.